internal/pkg/earpointer/state.go

```diff
@@ -0,0 +1,391 @@
+	if v, ok := fmap[getDirectPointToField()]; ok {
+		return "--> " + v.String()
+	}
+	var fstrs []string
+	for k, v := range fmap {
+		fstrs = append(fstrs, k.Name+"->"+v.String())
+	}
+	sort.Strings(fstrs)
+	return "[" + strings.Join(fstrs, ", ") + "]"
+}
+
+// Various state mutation operation
+
+// Insert inserts reference "aref" to the state and returns the current
+// partition representative of "aref".
+func (state *State) Insert(aref Reference) Reference {
+	if state.finalized {
+		log.Fatal("No mutations allowed after Finalize()")
+	}
+	// Lookup failure will create new entry in the parent table.
+	failureCallback := func(ref Reference) Reference {
+		state.parents[ref] = ref
+		state.partitions[ref] = PartitionInfo{numMembers: 0, fieldRefs: make(ReferenceFieldMap)}
+		return ref
+	}
+	return state.lookupPartitionRep(aref, failureCallback)
+}
+
+// Unify unifies the references "aref1" and "aref2"
+func (state *State) Unify(aref1 Reference, aref2 Reference) {
+	state.UnifyReps(state.GetPartitionRep(aref1), state.GetPartitionRep(aref2))
+}
+
+// UnifyReps unifies the references "aref1" and "aref2", assuming both are
+// partition representative. This allows to avoid calling GetPartitionRep
+// unnecesarilly. Using this function with non-representative
+// references will raise an error.
+func (state *State) UnifyReps(aref1 Reference, aref2 Reference) {
+	if state.finalized {
+		log.Fatal("No mutations allowed after Finalize()")
+	}
+	// If they are in the same partition already, we are done.
+	if aref1 == aref2 {
+		return
+	}
+	// Create local copies to use swap (this improves the readability).
+	prep1, prep2 := aref1, aref2
+	// Find partition-level info for "aref1" and "aref2"
+	pinfo1 := state.getPartitionInfo(prep1)
+	pinfo2 := state.getPartitionInfo(prep2)
+	// swap so that "aref1" is always the smaller partition
+	if pinfo1.numMembers > pinfo2.numMembers {
+		prep1, prep2 = prep2, prep1
+		pinfo1, pinfo2 = pinfo2, pinfo1
+	}
+
+	// Create state by having "prep1" point to "prep2" as parent and
+	// by erasing "prep2" as a partition rep. We then call MergeFieldMap()
+	// to merge field maps (which can trigger further unification). This process
+	// will converge since the number of partitions is guaranteed to decrease at
+	// every unification and is lower-bounded by 1.
+
+	// Update parent pointer and partition info map for aref1 while retaining
+	// its field map in fmap1 and number of members in n1
+	state.parents[prep1] = prep2
+	fmap1 := pinfo1.fieldRefs
+	pinfo1.fieldRefs = make(ReferenceFieldMap)
+	n1 := pinfo1.numMembers
+	delete(state.partitions, prep1)
+
+	// Update partition info for aref2 and merge the field map fmap1 in-place.
+	pinfo2.numMembers += n1
+	state.mergeFieldMap(fmap1, pinfo2.fieldRefs)
+}
+
+// Finalize indicates that no more mutation operations will be performed.
+// After this call, internal data structures are optimized for lookups only,
+// and no mutations are allowed.
+func (state *State) Finalize() {
```

Comment:
This looks like a strong candidate to split the typestate into two different types. What is the expected gain from keeping this together compared to a builder and a result type?

Reply:
Now the state is broken into two parts, "State" for the mutations (so as to act like a builder), and "PartitionState" after the heap has been fully constructed. The real builder is the pointer analysis pass, and the State here contains the foundational operations for unification and lookups. While it is possible to move the mutation operations to the pointer analysis pass, it is better to put some here since that pass is mainly about processing SSA elements.
"State" now makes most methods private, and a user of the analysis should use the PartitionState rather than State.